lib/amazonka/src/Amazonka/Env/Hooks.hs

```diff
@@ -0,0 +1,420 @@
+    Hook_ (Request a, ClientResponse (AWSResponse a))
+  ) ->
+  Hooks ->
+  Hooks
+responseHook f hooks@Hooks {response} =
+  hooks {response = f response}
+
+{-# INLINE errorHook #-}
+errorHook ::
+  ( forall a.
+    (AWSRequest a, Typeable a) =>
+    Hook_ (Request a, Error) ->
+    Hook_ (Request a, Error)
+  ) ->
+  Hooks ->
+  Hooks
+errorHook f hooks@Hooks {error} =
+  hooks {error = f error}
+
+-- | Turn a @'Hook' a@ into another @'Hook' a@ that does nothing.
+--
+-- -- Example: remove all request hooks:
+-- @
+-- requestHook noHook :: Hooks -> Hooks
+-- @
+noHook :: Hook a -> Hook a
+noHook _ _ = pure
+
+-- | Turn a @'Hook_' a@ into another @'Hook_' a@ that does nothing.
+--
+-- -- Example: Remove all response hooks:
+-- @
+-- responseHook noHook_ :: Hooks -> Hooks
+-- @
+noHook_ :: Hook_ a -> Hook_ a
+noHook_ _ _ _ = pure ()
+
+-- | When @a@ and @b@ are the same type, run the given 'Hook a' after
+-- all others, otherwise do nothing.
+--
+-- -- Example: Run @getObjectRequestHook@ on anything that is a @GetObjectRequest@:
+-- @
+-- requestHook (hookMatchingType @GetObjectRequest getObjectRequestHook) :: Hooks -> Hooks
+-- @
+hookMatchingType ::
+  forall a b. (Typeable a, Typeable b) => Hook a -> Hook b -> Hook b
+hookMatchingType newHook oldHook = \env ->
+  oldHook env >=> case eqT @a @b of
+    Just Refl -> newHook env
+    Nothing -> pure
+
+-- | When @a@ and @b@ are the same type, run the given 'Hook_ a' after
+-- all other hooks have run.
+--
+-- -- Example: Run @aSignedRequestHook@ on anything that is a @Signed GetObjectRequest@:
+-- @
+-- requestHook (hookMatchingType_ @(Signed GetObjectRequest) aSignedRequestHook) :: Hooks -> Hooks
+-- @
+hookMatchingType_ ::
+  forall a b. (Typeable a, Typeable b) => Hook_ a -> Hook_ b -> Hook_ b
+hookMatchingType_ newHook oldHook = \env a -> do
+  oldHook env a
+  case eqT @a @b of
+    Just Refl -> newHook env a
+    Nothing -> pure ()
+
+-- | When @a@ and @b@ are the same type, do not call any more hooks.
+--
+-- -- Example: Prevent any request hooks from running against a @PutObjectRequest@:
+-- @
+-- requestHook (blockMatchingType @PutObjectRequest) :: Hooks -> Hooks
+-- @
+blockMatchingType :: forall a b. (Typeable a, Typeable b) => Hook b -> Hook b
+blockMatchingType oldHook = \env -> case eqT @a @b of
+  Just Refl -> pure
+  Nothing -> oldHook env
+
+-- | When @a@ and @b@ are the same type, do not call any more hooks.
+--
+-- -- Example: Prevent any error hooks from running against errors caused by a @PutObjectRequest@:
+-- @
+-- errorHook (blockMatchingType @(Request PutObjectRequest, Error)) :: Hooks -> Hooks
+-- @
+blockMatchingType_ :: forall a b. (Typeable a, Typeable b) => Hook_ b -> Hook_ b
+blockMatchingType_ oldHook = \env a -> case eqT @a @b of
+  Just Refl -> pure ()
+  Nothing -> oldHook env a
+
+-- | Run the wrapped hook unless the given 'Fold' or 'Traversal'
+-- matches the error. With the error matches in each service binding,
```

Comment:
Something wrong with this sentence, not sure what it was supposed to say.

Reply:
It's a pointer to look up each service's error _matchers_. Will reword.